### PR TITLE
Log batch write for asyncAtomicOp used scan->insert->delete

### PR DESCRIPTION
**Pull request by bright-starry-sky**

### CODE REVIEW

File: src/kvstore/LogEncoder.cpp [hunk 1]
```diff
@@ -164,6 +164,64 @@ std::vector<folly::StringPiece> decodeMultiValues(folly::StringPiece encoded) {
...
+    auto num = static_cast<uint32_t>(batch.size());
+    encoded.append(reinterpret_cast<char*>(&num), sizeof(uint32_t));
+    // Values
+    encoded.append(data.c_str(), data.size());
+
+    return encoded;
+}
+
+std::vector<std::pair<LogType, std::pair<std::string, std::string>>>
+decodeBatchValuee(folly::StringPiece encoded) {
```
Comment (critical27) [inline src/kvstore/LogEncoder.cpp]: typo decodeBatchValuee

Comment (bright-starry-sky) [reply src/kvstore/LogEncoder.cpp]: Let me correct it , thanks

File: src/kvstore/LogEncoder.h [hunk 2]
```diff
@@ -19,10 +19,11 @@ enum LogType : char {
...
     OP_REMOVE_PREFIX  = 0x5,
     OP_REMOVE_RANGE   = 0x6,
-    OP_ADD_LEARNER    = 0x07,
-    OP_TRANS_LEADER   = 0x08,
-    OP_ADD_PEER       = 0x09,
-    OP_REMOVE_PEER    = 0x10,
+    OP_BATCH_WRITE    = 0x7,
+    OP_ADD_LEARNER    = 0x8,
+    OP_TRANS_LEADER   = 0x9,
+    OP_ADD_PEER       = 0x10,
```
Comment (wadeliuyi) [inline src/kvstore/LogEncoder.h]: 0xA? and 0xB?

Comment (bright-starry-sky) [reply src/kvstore/LogEncoder.h]: Wow，excellent ！Let me take care of it. thanks.

**Comment by bright-starry-sky** [comment]
Resolved wadeliuyi's comment

File: src/kvstore/LogEncoder.cpp [hunk 3]
```diff
@@ -164,6 +164,64 @@ std::vector<folly::StringPiece> decodeMultiValues(folly::StringPiece encoded) {
...
+        p += sizeof(LogType);
+        uint32_t len1 = *(reinterpret_cast<const uint32_t*>(p));
+        p += sizeof(uint32_t);
+        auto v1 = folly::StringPiece(p, len1);
+        p += len1;
+        uint32_t len2 = *(reinterpret_cast<const uint32_t*>(p));
+        p += sizeof(uint32_t);
+        auto v2 = folly::StringPiece(p, len2);
+        p += len2;
+        std::pair<std::string, std::string> kv(v1.str(), v2.str());
```
Comment (dangleptr) [inline src/kvstore/LogEncoder.cpp]: Could we avoid the string copy here?

Comment (bright-starry-sky) [reply src/kvstore/LogEncoder.cpp]: Good point , let me take a look .

**Comment by bright-starry-sky** [comment]
Address @dangleptr's comment.

File: src/kvstore/LogEncoder.h [hunk 4]
```diff
@@ -40,12 +47,49 @@ std::string encodeMultiValues(LogType type,
...
+std::vector<std::pair<BatchLogType, std::pair<std::string, std::string>>>
+decodeBatchValue(folly::StringPiece encoded);
 
 std::string encodeHost(LogType type, const HostAddr& learner);
 HostAddr decodeHost(LogType type, const folly::StringPiece& encoded);
 
 int64_t getTimestamp(const folly::StringPiece& command);
 
+
+class BatchLogHelper {
```
Comment (darionyaphet) [inline src/kvstore/LogEncoder.h]: Generally LGTM 🎉  Personally think `BatchHolder` would be better？

Comment (bright-starry-sky) [reply src/kvstore/LogEncoder.h]: Good idea, let me rename it.

File: src/kvstore/LogEncoder.h [hunk 5]
```diff
@@ -19,10 +19,17 @@ enum LogType : char {
     OP_MULTI_REMOVE   = 0x4,
     OP_REMOVE_PREFIX  = 0x5,
     OP_REMOVE_RANGE   = 0x6,
-    OP_ADD_LEARNER    = 0x07,
-    OP_TRANS_LEADER   = 0x08,
-    OP_ADD_PEER       = 0x09,
-    OP_REMOVE_PEER    = 0x10,
+    OP_BATCH_WRITE    = 0x7,
```
Comment (dangleptr) [inline src/kvstore/LogEncoder.h]: Don't change the value.  It is incompatible for old WALs

File: src/kvstore/LogEncoder.h [hunk 6]
```diff
@@ -40,12 +47,49 @@ std::string encodeMultiValues(LogType type,
...
 
 int64_t getTimestamp(const folly::StringPiece& command);
 
+
+class BatchHolder {
+public:
+    BatchHolder() = default;
+    ~BatchHolder() = default;
+
+    void put(const std::string& key, const std::string& val) {
```
Comment (dangleptr) [inline src/kvstore/LogEncoder.h]: Could we take right reference here to avoid extra copy.

File: src/kvstore/LogEncoder.cpp [hunk 7]
```diff
@@ -164,6 +164,63 @@ std::vector<folly::StringPiece> decodeMultiValues(folly::StringPiece encoded) {
     return values;
 }
 
+std::string encodeBatchValue(const std::vector<std::pair<BatchLogType,
+                             std::pair<std::string, std::string>>>& batch) {
+    auto type = LogType::OP_BATCH_WRITE;
+    std::string data, encoded;
```
Comment (dangleptr) [inline src/kvstore/LogEncoder.cpp]: Why use data here?  just encoded is enough.
Before append data into "encoded",  please reserve enough space for it.

**Comment by bright-starry-sky** [comment]
To address @dangleptr's comment.

File: src/kvstore/LogEncoder.cpp [hunk 8]
```diff
@@ -164,6 +164,61 @@ std::vector<folly::StringPiece> decodeMultiValues(folly::StringPiece encoded) {
...
+        encoded.append(v1.c_str(), len);
+        len = v2.size();
+        encoded.append(reinterpret_cast<char*>(&len), sizeof(uint32_t));
+        encoded.append(v2.c_str(), len);
+    }
+
+    return encoded;
+}
+
+std::vector<std::pair<BatchLogType, std::pair<std::string, std::string>>>
```
Comment (dangleptr) [inline src/kvstore/LogEncoder.cpp]: Could we return std::pair<folly::stringPiece,  folly::stringPiece>

File: src/kvstore/LogEncoder.cpp [hunk 9]
```diff
@@ -164,6 +164,61 @@ std::vector<folly::StringPiece> decodeMultiValues(folly::StringPiece encoded) {
...
+    for (auto i = 0U; i < numValues; i++) {
+        auto offset = 0;
+        BatchLogType type = *(reinterpret_cast<const BatchLogType *>(p));
+        p += sizeof(LogType);
+        uint32_t len1 = *(reinterpret_cast<const uint32_t*>(p));
+        offset += sizeof(uint32_t) + len1;
+        uint32_t len2 = *(reinterpret_cast<const uint32_t*>(p + offset));
+        offset += sizeof(uint32_t);
+        batch.emplace_back(type,
+                std::make_pair(folly::StringPiece(p + sizeof(uint32_t), len1).str(),
```
Comment (dangleptr) [inline src/kvstore/LogEncoder.cpp]: We could avoid the copy.

File: src/kvstore/LogEncoder.cpp (hunk 8 above)
Comment (bright-starry-sky) [reply src/kvstore/LogEncoder.cpp]: Wow, I had an epiphany that original return type will be memory copy occur.
Good cache. and good prompt.

**Pull request merged**